### DISCUSSION TITLE
File Upload field issue

### DISCUSSION THREAD

**Discussion by AnatolyKruglov** [Laravel-Backpack/community-forum#286] [Q&A (Help)] [answered]
I'm trying to upload images using backpack's File Upload field and store them in the /uploads folder. The field appears in the admin interface, allows me to browse files, but the chosen image isn't uploaded to the server and I don't get any errors. 

I followed the official documentation thoroughly + applied advice found on forums (none of which worked), so I assume this might be a bug(

In my CrudController:
```php
    CRUD::addField([   
            'name'      => 'image',
            'label'     => 'Image',
            'type'      => 'upload',
            'upload'    => true,
            'disk'      => 'uploads', 
    ]);
```

In my Model:
```php
    public function setImageAttribute($value)
    {
        $attribute_name = "image";
        $disk = "uploads";
        $destination_path = ""; //relative to $disk
        $this->uploadFileToDisk($value, $attribute_name, $disk, $destination_path);
    }
```

In `config/filesystems.php`:
```php
    'disks' => [
        'uploads' => [
            'driver' => 'local',
            'root' => env('APP_URL') . '/public_html/uploads',   //btw, my public folder is called public_html
            'url' => env('APP_URL') . '/uploads',
            'visibility' => 'public',
        ],
     ...
```

I would appreciate any help with that)

**Comment by welcome** [comment]
Hello there! Thanks for opening your first issue on this repo!

Just a heads-up: **Here at Backpack we use Github Issues only for tracking bugs**. Talk about new features is also acceptable. This helps _a lot_ in keeping our focus on improving Backpack. **If you issue is not a bug/feature, please help us out by closing the issue yourself and posting in the appropriate medium (see below)**. If you're not sure where it fits, it's ok, a community member will probably reply to help you with that.

Backpack communication channels:
 - Bug Reports, Feature Requests - Github Issues (here);
 - Quick help (_How do I do X_) - [Gitter Chatroom](https://gitter.im/BackpackForLaravel/Lobby);
 - Long questions (_I have done X and Y and it won't do Z wtf_) - [Stackoverflow](https://stackoverflow.com/questions/tagged/backpack-for-laravel), using the ```backpack-for-laravel``` tag;
 - Showing off something you've made, asking for opinion on Backpack/Laravel matters - [Reddit](https://www.reddit.com/r/BackpackForLaravel/);

Please keep in mind Backpack offers no official / paid support. Whatever help you receive here, on Gitter, Slack or Stackoverflow is thanks to our awesome _awesome_ community members, who give up some of their time to help their peers. If you want to join our community, just start pitching in. We take pride in being a welcoming bunch.

Thank you!

--
Justin Case
The Backpack Robot

**Comment by tabacitu** [comment]
I've just tested this and it worked fine - so I'm pretty sure it's not a Backpack bug, but an implementation issue. I can't spot anything obvious from your code... I'd recommend you go one-by-one in the lines above, `dd()` the stuff you use to determine where the problem lies.

If the field is visible, no need to test that. So I'd start with the Model, inside the `setImageAttribute()` mutator. `dd($value)` then if that's ok copy-paste the contents of `uploadFileToDisk()` to your model and `dd()` along that one too.

99% of the times this is an issue with how the filesystem is configured though. Try it the same way it's done inside [our demo](https://github.com/laravel-backpack/demo) too (where it's 100% working). In fact, if you end up with nothing, you can install the demo too, to see if it works as-is (it should).

Hope it helps.
Cheers!

**Comment by mbc-audiovisuales** [comment]
Hi,

using dd(request()) shows me that no file is coming to be taken by 'uploadFileToDisk' function to store, so DB is correctly updated while no file is written to disk as CRUD 'upload' field seems to bring no file at all to server side when trying to create/update a model with that 'upload' field populated with any local file. Weird

**Comment by pxpm** [comment]
Not sure but might be because file uploads are disabled in demo?

Did you tried in a new laravel installation with backpack?

Let me know,
Pedro

A quinta, 29/07/2021, 12:07, MBC Servicios Audiovisuales <
***@***.***> escreveu:

**Reply by mbc-audiovisuales** [reply]
Hi, I managed to discover the problem.

CRUD panel's form does not include the attribute `enctype="multipart/form-data"`.

I guess something is missing in docs regarding CRUD Controller setup around the usage of 'upload' field. As shown into source code for the blade files create.blade.php, edit.blade.php, for backpack/crud, I can see a conditional inclusion of this attribute as follows (for the edit.blade.php case, similarly found into create.blade.php template):
```
@if ($crud->hasUploadFields('update', $entry->getKey()))
enctype="multipart/form-data"
@endif
```

Looking for that `hasUploadFields` I discovered that 'upload' CRUD field shall include `'upload' => true` attribute to work as expected. Neat coding may need that attribute to be defaulted to true in case it is not provided.

Cheers!

**Comment by jcastroa87** [comment]
Hi.

Due to not activity on many years i will close this discussion, if there is needed feel free to reopen or create a new one.

Cheers.

**Comment by pxpm** [comment, marked as answer]
Also as of now (Backpack 6) you can use the https://backpackforlaravel.com/docs/6.x/crud-uploaders to make the uploading of files a breeze in Backpack. 

Cheers
